### DISCUSSION TITLE
How to run the simulation on local machine to test my code

### DISCUSSION THREAD

**Discussion by rua0ra1** [cyberbotics/wrestling#21] [Q&A] [answered]
Hi,
My name is Ram and I am participating in the Humanoid Competition. I am quite new to the we-bot simulation. I am wondering how can I run the simulation on my local machine to test my code. Currently, I am pushing my changes to Github and analysing the simulation result in webot competition page. I feel its not a idle way to test my code. I really appreciate it if you can help with this issue.

**Comment by omichel** [comment, marked as answer]
You should download and install [Webots R2023a](https://github.com/cyberbotics/webots/releases) on your local computer. I would recommend that you install R2023a-rev1 from a recent nightly build as it fixes some problems with Python controllers. Then, you can `git clone` on your local computer your repository with your code for the competition. Finally, you can open the `wrestling.wbt` world file from the `worlds` folder of your repository cloned on your local computer. You will then be able to run the simulation locally, make changes to your robot controller, test them locally and `git commit` them to the main branch of your repository. They will be run against the other participants only after you `git push` them on GitHub.

**Note:** if you are not familiar with git, I would recommend you to follow some introductory tutorial about it.

**Reply by rua0ra1** [reply]
Thank you for the detail answer @omichel

**Reply by pemile29** [reply]
Thank you also for this help

After installing the webots-2023a package on a debian, the webots environment works fine but I can't use the Camera and ImageProcessing imports.

Error message at runtime on the import line:
import numpy as np
ModuleNotFoundError: No module named 'numpy'

My Dockerfile contains the FROM cyberbotics/webots.cloud:R2023a-ubuntu20.04-numpy and the RUN install

Did I miss a step?

**Reply by omichel** [reply]
When running locally, you should install the python modules on your local system, e.g., `pip install numpy`, etc.

**Reply by pemile29** [reply]
Thanks, everything works, sorry I'm a beginner in python and I discover webots.

I'll take advantage of your incredible availability to ask you where to find the "motion editor" that the documentation talks about?

**Reply by omichel** [reply]
Unfortunately, there is no motion editor any more.

**Reply by pemile29** [reply]
Bad news ! 

It seems to me an indispensable tool, no?

**Reply by omichel** [reply]
Not really.
1. You can edit these motion files manually (they are plain text).
2. You can use the motion library to make the robot walk.

**Reply by kalashjain23** [reply]
> When running locally, you should install the python modules on your local system, e.g., `pip install numpy`, etc.

I've installed all the required modules for my robot, but it still shows the same error:
```
import numpy as np
ModuleNotFoundError: No module named 'numpy'
```

When I'm running `pip install numpy`:
Requirement already satisfied: numpy in /usr/lib/python3/dist-packages (1.21.5)
  
Can you guide me through this one?

**Reply by Jean-Eudes-le-retour** [reply]
This doesn't seem like a Webots problem.

It looks like you might have multiple versions of python installed on your system, have a look at these links for pointers:
https://github.com/numpy/numpy/issues/15424
https://www.quora.com/What-is-Python-m-pip

**Comment by pemile29** [comment]
The motion file library is still very limited!

I used it to write a simple spinning robot (which was first until 30 minutes ago) but I guess it won't be enough for long !

Editing motion files by hand doesn't seem easy to me !

PS: Is it possible to use the GPS to avoid going out of the ring ?

**Reply by omichel** [reply]
No. There is no GPS on a real NAO robot. You should rely on vision and possibly odometry to avoid going out of the ring.

**Reply by pemile29** [reply]
I can't imagine how to use odometry that would take into account crashes and falls !

Using vision to locate oneself inside the ring seems beyond my capabilities with opencv :-/

Anyway, thanks for your help

**Reply by omichel** [reply]
Just some clues: using the second camera of the robot, which looks down, it might not be super difficult to locate on the camera image the red line delimiting the ring and to estimate how far the robot is from this line by measuring the apparent width of the line...

**Reply by pemile29** [reply]
Right, thank you for motivating me, although my reflex in front of opencv is to use TurnLeft180.motion !

**Reply by pemile29** [reply]
One more novice question, in the "Eve" demo what does the camera.send_to_robot_window() function do ?

And how can i display the working image tagged with the circle in the displayed stream of the video ?

PS: in my locally installed webots, the videos of the lower cameras are reversed on the screen :
On the left the high camera image is the participant's and the low camera is the opponent's and vice versa on the right !

**Reply by Jean-Eudes-le-retour** [reply]
> what does the camera.send_to_robot_window() function do ?

The `camera.send_to_robot_window(img)` takes an OpenCV `img` argument and formats it to send it to a compatible robot window.

I forgot to add a working robot window to Eve's files, you can now pull the latest changes in [Eve's repository](https://github.com/cyberbotics/wrestling-eve) and you should see a new window pop up when you open the world file. You can then go to the "Image feed" tab and when the controller will run you will see the image processing result on top of the robot's camera images.
